rust/protocol/src/address.rs

```diff
@@ -1,26 +1,57 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+//! Where to send a [sealed_sender] message to.
+
+// These are used in docstrings.
+#[allow(unused_imports)]
+use crate::sealed_sender::{self, UnidentifiedSenderMessageContent};
+
 use std::fmt;
 
+/// The type used in memory to represent a *device*, i.e. a particular Signal client instance which
+/// represents some user.
+///
+/// Used in [ProtocolAddress].
+pub type DeviceId = u32;
+
+/// The target of an [UnidentifiedSenderMessageContent].
 #[derive(Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
 pub struct ProtocolAddress {
     name: String,
-    device_id: u32,
+    device_id: DeviceId,
 }
 
 impl ProtocolAddress {
-    pub fn new(name: String, device_id: u32) -> Self {
+    /// Create a new address.
+    ///
+    /// - `name` defines a user's public identity, and therefore must be globally unique to that
+    /// user.
+    /// - Each Signal client instance then has its own `device_id`, which must be unique among
+    ///   all clients for that user.
+    pub fn new(name: String, device_id: DeviceId) -> Self {
         ProtocolAddress { name, device_id }
     }
 
+    /// A unique identifier for the target user.
+    ///
+    /// In the Signal app, this is currently a phone number.
```

Comment:
This is not true; it's usually a UUID (and that may be enforced at some point in the future). ProtocolAddress still accepts a free-form string so that it works with phone numbers, but that's mostly just to support tests. All the actual app logic should be using UUIDs.

Reply:
AWESOME!!!!!!!!! might tweet about this

Reply:
https://twitter.com/hipsterelectron/status/1389639068862930944